File: src/levels/L05-adapter/adapter.rb
class Cashier
  def initialize(gateway, barista)
    @gateway = gateway
    @barista = barista
  end

  # Cashier no cambió: sigue hablando su idioma.
  # region: Cashier#charge
  def charge(order)
    @gateway.charge(order.total)
    @barista.prepare(order)
  end
  # endregion
end

# Traduce la interfaz que Cashier espera a la que el SDK ofrece.
# region: PagoFacilAdapter#charge
class PagoFacilAdapter
  def initialize(sdk) = @sdk = sdk

  def charge(total)
    @sdk.cobrar_en_centavos(monto_centavos: (total * 100).round, referencia: SecureRandom.uuid)
  end
end
# endregion

File: src/levels/L05-adapter/ts/adapter.ts
interface PaymentGateway {
  charge(total: number): void
}

class Cashier {
  constructor(
    private gateway: PaymentGateway,
    private barista: Barista,
  ) {}

  // Cashier no cambió: sigue hablando su idioma.
  // region: Cashier#charge
  charge(order: Order) {
    this.gateway.charge(order.total)
    this.barista.prepare(order)
  }
  // endregion
}

// Traduce la interfaz que Cashier espera a la que el SDK ofrece.
// region: PagoFacilAdapter#charge
class PagoFacilAdapter implements PaymentGateway {
  constructor(private sdk: PagoFacil.SDK) {}

  charge(total: number) {
    this.sdk.cobrarEnCentavos({ montoCentavos: Math.round(total * 100), referencia: crypto.randomUUID() })
  }
}
// endregion
